Realm/RLMProperty.mm

```diff
@@ -343,6 +343,9 @@ - (instancetype)initSwiftPropertyWithName:(NSString *)name
     }
     else if ([_objcRawType isEqualToString:@"@\"NSNumber\""]) {
         const char *numberType = [propertyValue objCType];
+        if (!numberType) {
+            @throw RLMException(@"Can't persist NSNumber directly: use a Swift-native number type.");
```

Comment:
The Realm Objective-C docs don't mention support for storing non-nullable `NSNumber`s—the primitive type should be used—so we could at least throw an exception for the non-`Optional` case. This would not have helped #3705 though.